### ISSUE TITLE
terraform plan -light

### ISSUE THREAD

**Issue by ricardbejarano** [hashicorp/terraform#35290]
### Terraform Version

```shell
Terraform v1.8.4
on darwin_amd64
```

### Use Cases

Fixing plan (or rather, refresh) performance on large enough (in resource count) states.

### Attempted Solutions

- Increasing parallelism only takes you so far until the providers rate-limit you.
- Disabling refresh works but comes with all sorts of consistency risks.
- Splitting one big state into multiple smaller states comes with all sorts of problems associated with it.

### Proposal

Add a `terraform plan -light` flag such that only resources modified in code are targeted for planning.

This would reduce the scope of the pre-plan refresh down to the set of resources we know changed, which reduces overall plan times without the consistency risk of `-refresh=false`.

In order for Terraform to know what resources were modified in code, it would store a copy of the code in state every time it successfully applies. This would allow diff'ing "last-applied code" vs. "new code", the result of which is the scope of the next "light" plan.

Basically, `-light` tells Terraform to autogenerate the `-target` list from code changes.

More details: https://www.bejarano.io/terraform-plan-light/

**Comment by crw** [comment]
Thanks for this feature request! If you are viewing this issue and would like to indicate your interest, please use the 👍 reaction on the issue description to upvote this issue. We also welcome additional use case descriptions. Thanks again!

**Comment by Tbohunek** [comment]
### THIS IS BIG! 🔝🚀🔥 Thanks @ricardbejarano! ###
Make Terraform faster and save CO2? Every Terraform user wants this, whether they know it or not. And I'll do all I can to propagate this idea across my organization and to anyone I bump into, including my mom, who now has a GitHub account! :octocat:

#### Adding the benefits that I see: ####
- **Reduce concurrency clashes**
_Other plans cause locks, queues, or API depletion; other applies cause me to have to replan again._
- **Reduce the likelihood of external changes breaking your change**
_Unrelated `data.azuread_user` fails the whole `plan` because the user no longer exists._
- **Reduce the likelihood of unrelated API problems breaking your change**
_Just a week ago, the `azurerm_security_center_contact` API was down, making TFE `apply` impossible._
- **Reduce Provider API limits exhaustion**, especially in hub and spoke topologies
_`azurerm` and `google` providers often time out reads with API limits of ~25 reads/sec._
- **Reduce CO2 emissions and compute resource waste**
_Eating beef is being denounced while Terraform literally wastes CO2 and energy with pointless reads._
- **Make VCS-driven TFE workflow feasible again**
_VCS-driven TFE workflow cannot `apply -target=what.you.want` to circumvent the above issues._

If this gets implemented, **every single TFE user will be grateful, knowingly or unknowingly**.
When released, make them aware through a tooltip with credits. 😉

**Comment by jsimoni** [comment]
something similar to https://nx.dev/ci/features/affected ?

**Comment by igtsekov** [comment]
This idea can be extended to `init` as well.

**Comment by ricardbejarano** [comment]
BTW I've been doing [this](https://github.com/opentofu/opentofu/issues/1703#issuecomment-2198216730) for a few months now and it is fantastic, if you want to have this feature before it gets developed:

```bash
terraform plan $(terraform plan -refresh=false -json | jq -r '.change.resource.addr|select(.!=null)|"-target="+.')
```

- `.change.resource.addr` is where the full resource addresses are in the JSON-formatted output
- `select(.!=null)` drops all null values (output lines that didn't contain the previous)
- `"-target="+.` prepends the `-target=` prefix to the resulting lines without the need to call `sed`, `xargs` or anything else

**Comment by ricardbejarano** [comment]
@jbardin @crw may I ask what's next for this? Shall I write the PR?

**Comment by crw** [comment]
@ricardbejarano thanks for the offer, however it is unlikely to be reviewed until the maintainers have considered the larger ramifications of such a change. There is interest in exploring this at some point in the future, I will keep you posted.

**Comment by ricardbejarano** [comment]
@austinvalle awesome!

**Comment by antonbabenko** [comment]
@ricardbejarano Thank you for the inspiration!

**Comment by austinvalle** [comment]
Hi everyone 👋🏻, this feature is slated to release with `v1.17.0` (sometime in October '26)  and is available via the new plan flag `terraform plan -minimal-refresh` in the latest alpha release ([`v1.17.0-alpha20260827`](https://github.com/hashicorp/terraform/releases/tag/v1.17.0-alpha20260827)) if you're interested in trying it out.

Using `-minimal-refresh` is similar in principle to https://github.com/hashicorp/terraform/issues/35290#issuecomment-2376343446, per resource instance, so Terraform will first plan using the pre-refreshed state of a resource and if the change produced is a no-op, then the refresh will be skipped. If that plan does produce a change (anything but no-op), then Terraform will refresh the state for just that resource and run another plan. 

-------

I want to also highlight that `-minimal-refresh` is a different solution/behavior then what was originally proposed with [storing a hash of the configuration](https://www.bejarano.io/terraform-plan-light/). So with `-minimal-refresh`, if the state had drifted + then was refreshed by a separate process (say periodically refreshing state with `terraform refresh`), then `terraform plan -minimal-refresh` could still detect a change produced entirely by drift (it just would be coming from a command that was run earlier!)

I think this still provides a solid middle ground that should help improve performance while keeping all of the necessary systems of Terraform (namely, the providers!) in the loop for a plan. Doing this also means that providers can still influence the plan to prevent unnecessary refreshes when the configuration _does_ change, like if you added whitespace/changed ordering of a JSON string, but the provider knows that the remote system doesn't need to updated. With this new flag, the refresh would still be correctly skipped if the provider indicates that no change is necessary. 👍🏻 

-------

Let us know if you run into any issues with the new flag + thanks again to @ricardbejarano for the original idea! 🎉
